src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/CompareSymbolsCorrectlyAnalyzer.cs

```diff
@@ -38,15 +44,61 @@ public override void Initialize(AnalysisContext context)
             {
                 var compilation = context.Compilation;
                 var symbolType = compilation.GetTypeByMetadataName(s_symbolTypeFullName);
+
                 if (symbolType is null)
                 {
                     return;
                 }
 
                 context.RegisterOperationAction(context => HandleBinaryOperator(in context, symbolType), OperationKind.BinaryOperator);
+
+                var comparerTypes = s_comparerTypeFullName
+                    .Select(comparerTypeFullName => compilation.GetTypeByMetadataName(comparerTypeFullName))
+                    .WhereNotNull()
+                    .ToImmutableArray();
+
+                if (comparerTypes.IsDefaultOrEmpty)
+                {
+                    return;
+                }
+
+                context.RegisterOperationAction(context => HandleOperation(in context, comparerTypes), OperationKind.ObjectCreation);
```

Comment:
I don't think a constructor check is good enough here. For example: https://docs.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutabledictionary.createbuilder?view=netcore-2.2, we need to ensure that we're calling the comparer overload of this static method. I'm not sure what a good heuristic is, but some more thought needs to be put into other things that operate on equality.

Reply:
That seems more difficult, and I don't think should block this. Thanks for pointing it out though. 

We can probably whitelist some cases, but I don't know of a generic rule that would make sense. For any type `T`, static methods that return `T` _could_ set a comparison without it being specific. If it's an API type we could probably formulate a known list... for other types, theoretically it gets covered by the constructor operator? 

Known lists gets another maintenance cost though.

Reply:
I agree that complicated heuristics might be too tricky here (but really not much worse than a constructor heuristic), but Roslyn itself totally has ImmutableDictionaries with the key being a symbol. So we can't ignore this case.